Competition2017/src/competition/subsystems/drive/commands/FieldOrientedTankDriveWithJoystick.java

```diff
@@ -0,0 +1,51 @@
+package competition.subsystems.drive.commands;
+
+import java.util.function.DoubleSupplier;
+
+import com.google.inject.Inject;
+
+import competition.operator_interface.OperatorInterface;
+import competition.subsystems.drive.DriveSubsystem;
+import competition.subsystems.pose.PoseSubsystem;
+import xbot.common.controls.sensors.XJoystick;
+import xbot.common.math.PIDFactory;
+import xbot.common.math.PIDManager;
+
+public class FieldOrientedTankDriveWithJoystick extends BaseDriveCommand {
+
+    private final PoseSubsystem poseSubsystem;
+    
+    private final XJoystick leftJoystick;
+    
+    private final PIDManager headingDrivePid;
+    public final double defaultPValue = 1 / 80d;
+
+    protected DoubleSupplier targetHeadingSupplier;
+    
+    @Inject
+    public FieldOrientedTankDriveWithJoystick(
+            OperatorInterface oi, 
+            DriveSubsystem driveSubsystem, 
+            PoseSubsystem poseSubsystem,
+            PIDFactory pidFactory) {
+        super(driveSubsystem);
+        this.poseSubsystem = poseSubsystem;
+        this.leftJoystick = oi.leftJoystick;
+        headingDrivePid = pidFactory.createPIDManager("Rotate to heading", defaultPValue, 0, 0, 0, 1, -1, 3, 3.0 / 20, 1);
+    }
+
+    @Override
+    public void initialize() {
+        log.info("Initializing");
+        targetHeadingSupplier = () -> leftJoystick.getVector().getAngle();
+    }
+
+    @Override
+    public void execute() {
+        double translationPower = leftJoystick.getVector().getMagnitude();
+        
+        double headingError =  poseSubsystem.getCurrentHeading().difference(targetHeadingSupplier.getAsDouble());
+        double rotationPower = -headingDrivePid.calculate(0, headingError);
```

Comment:
This algorithm will drive the robot in a spiral while it rotates, which seems undesirable. Can you add some math that incorporates the absolute heading difference into the translation power? I'm thinking that the robot shouldn't start driving forward at all until it's within, say, 25 degrees; you could have the power grow linearly or exponentially after that.

Also, there should be a deadband on the deflection of the stick such that the robot doesn't attempt to rotate until you've deflected the stick past a certain threshold.